praw/models/reddit/emoji.py

```diff
@@ -0,0 +1,71 @@
+"""Provide the EMOJI class."""
+from ...const import API_PATH
+from .base import RedditBase
+
+
+class Emoji(RedditBase):
+    """An individual Emoji object."""
+
+    __hash__ = RedditBase.__hash__
+    STR_FIELD = 'name'
+
+    def __init__(self, reddit, subreddit, name, _data=None):
+        """Construct an instance of the Emoji object."""
+        self.reddit = reddit
+        self.subreddit = subreddit
+        self.name = name
+        super(Emoji, self).__init__(reddit, _data)
+
+    def add(self, filepath):
+        """Add an emoji to this subreddit.
+
+        :param emoji: An emoji name (e.g., ``'cake'``) or
+            :class:`~.Emoji` instance.
+
+        To add ``'cake'`` to the subreddit ``'praw_test'`` try:
+
+        .. code:: python
+
+           reddit.subreddit('praw_test').emoji['cake'].add('cake.png')
+
+        """
+        data = None
+        if filepath.lower().endswith('.png'):
+            data = {'filepath': filepath, 'mimetype': 'image/png'}
+        elif filepath.lower().endswith('.jpg') \
+                or filepath.lower().endswith('.jpeg'):
+            data = {'filepath': filepath, 'mimetype': 'image/jpeg'}
+        if data is not None:
+            url = API_PATH['emoji_lease'].format(
+                subreddit=self.subreddit, method='add')
+            s3_lease = self._reddit.post(url, data=data)['s3UploadLease']
+            s3_url = 'https:' + s3_lease['action']
+            # get a raw requests.Session to contact non-reddit domain
+            http = self.subreddit._reddit._core._requestor._http
+            s3_parameters = dict((item['name'], item['value'])
+                                 for item in s3_lease['fields'])
+            with open(filepath, 'rb') as file:
+                s3_parameters['file'] = file
+                http.post(s3_url, files=s3_parameters)
```

Comment:
Two things. First what happens if this request fails? We want to ensure it succeeds.

Second, while it appears to work, the `s3_parameters` should not be part of the `files` and rather to data (perhaps rename to s3_data). Consider:

    response = http.post(s3_url, data=s3_data, files={'file': fp})
    response.raise_for_status()

At some point it would be good to wrap any possible request exceptions with the appropriate prawcore exceptions, but we'll probably need to extend prawcore so that it handles this request directly.